File: src/main/java/no/microdata/datastore/adapters/api/MetadataAPI.java
package no.microdata.datastore.adapters.api;

import no.microdata.datastore.MetadataService;
import no.microdata.datastore.model.MetadataQuery;
import no.microdata.datastore.transformations.VersionUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import javax.servlet.http.HttpServletResponse;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static no.microdata.datastore.adapters.api.Constants.*;
import static no.microdata.datastore.adapters.api.RequestId.verifyAndUpdateRequestId;

@RestController
@RequestMapping(produces = {"application/json;charset=UTF-8", "application/x-msgpack"})
class MetadataAPI {

    private final static Logger log = LoggerFactory.getLogger(MetadataAPI.class);

    @Autowired
    MetadataService metadataService;

    @RequestMapping(value = "/metadata/data-store", method = RequestMethod.GET)
    Map getDataStore(@RequestHeader(value = X_REQUEST_ID, required = false) String requestId,
                           @RequestHeader(value = ACCEPT_LANGUAGE, required = false) String language,
                           HttpServletResponse response) {

        log.info("Entering getDataStore() with requestId = {} and language {}", requestId, language);

        var verifiedRequestId = verifyAndUpdateRequestId(requestId);

        response.setHeader(X_REQUEST_ID, verifiedRequestId);
        response.setHeader(CONTENT_LANGUAGE, "no");

        return metadataService.findAllDataStoreVersions(verifiedRequestId);
    }

    @RequestMapping(value = "/metadata/data-structures", method = RequestMethod.GET)
    List<Map<String, Object>> getDataStructures(@RequestHeader(value = X_REQUEST_ID, required = false) String requestId,
                                @RequestHeader(value = ACCEPT_LANGUAGE, required = false) List<String> languages,
                                @RequestParam(value = NAMES, required = false) List names,
                                @RequestParam(value = VERSION, required = true) String version,
                                HttpServletResponse response) {

        MetadataQuery query = new MetadataQuery(
                new HashMap() {{
                    put("names", names);
                    put("languages", joinToString(languages));
                    put("requestId", verifyAndUpdateRequestId(requestId));
                    put("version", VersionUtils.toThreeLabelsIfNotDraft(version));
                }});

        log.info("Entering getDataStructures() where query = {}", query);

        response.setHeader(X_REQUEST_ID, query.getRequestId());
        response.setHeader(CONTENT_LANGUAGE, "no");

        return metadataService.findDataStructures(query);
    }

    @RequestMapping(value = "/metadata/all", method = RequestMethod.GET)
    Map<String, Object> getAllMetadata(@RequestHeader(value = X_REQUEST_ID, required = false) String requestId,
                       @RequestHeader(value = ACCEPT_LANGUAGE, required = false) List<String> languages,
                       @RequestParam(value = VERSION, required = true) String version,
                       HttpServletResponse response) {

        MetadataQuery query = new MetadataQuery(
                new HashMap() {{
                    put("languages", joinToString(languages));
                    put("requestId", verifyAndUpdateRequestId(requestId));
                    put("version", VersionUtils.toThreeLabelsIfNotDraft(version));
                }});

        log.info("Entering getAllMetadata() where query = {}", query);

        response.setHeader(X_REQUEST_ID, query.getRequestId());
        response.setHeader(CONTENT_LANGUAGE, "no");

        return metadataService.findAllMetadata(query);
    }

    @RequestMapping(value = "/languages", method = RequestMethod.GET)
    List<Map<String, Object>> getLanguages(@RequestHeader(value = X_REQUEST_ID, required = false) String requestId,
                           HttpServletResponse response) {

        log.info("Entering getLanguages() with request id = {}", requestId);

        requestId = verifyAndUpdateRequestId(requestId);
        response.setHeader(X_REQUEST_ID, requestId);

        return metadataService.findLanguages(requestId);
    }

    String joinToString(List<String> languages){
        return languages != null ? String.join(", ", languages) : null;
    }
}

File: metadata_service/api/metadata_api.py
import logging

from fastapi import APIRouter, Depends

from metadata_service.api.request_models import NameParam, MetadataQuery
from metadata_service.domain import metadata
from metadata_service.domain.version import get_version_from_string

logger = logging.getLogger()
metadata_router = APIRouter()


@metadata_router.get("/metadata/data-store")
def get_data_store():
    logger.info("GET /metadata/data-store")
    return metadata.find_all_datastore_versions()


@metadata_router.get("/metadata/data-structures/status")
def get_data_structure_current_status(validated_query: NameParam = Depends()):
    logger.info(
        f"GET /metadata/data-structures/status with name = {validated_query.names}"
    )
    return metadata.find_current_data_structure_status(
        validated_query.get_names_as_list()
    )


@metadata_router.post("/metadata/data-structures/status")
def get_data_structure_current_status_as_post(validated_body: NameParam):
    logger.info(
        f"POST /metadata/data-structures/status with name = {validated_body.names}"
    )
    return metadata.find_current_data_structure_status(
        validated_body.get_names_as_list()
    )


@metadata_router.get("/metadata/data-structures")
def get_data_structures(
    validated_query: MetadataQuery = Depends(),
):
    validated_query.include_attributes = True
    logger.info(f"GET /metadata/data-structures with query: {validated_query}")
    return metadata.find_data_structures(
        validated_query.names_as_list(),
        get_version_from_string(validated_query.version),
        validated_query.include_attributes,
        validated_query.skip_code_lists,
    )


@metadata_router.get("/metadata/all-data-structures")
def get_all_data_structures_ever():
    logger.info("GET /metadata/all-data-structures")
    return metadata.find_all_data_structures_ever()


@metadata_router.get("/metadata/all")
def get_all_metadata(
    validated_query: MetadataQuery = Depends(),
):
    logger.info(f"GET /metadata/all with version: {validated_query.version}")
    return metadata.find_all_metadata(
        get_version_from_string(validated_query.version),
        validated_query.skip_code_lists,
    )


@metadata_router.get("/languages")
def get_languages():
    logger.info("GET /languages")
    return metadata.find_languages()
